### PR TITLE
#22: Added validation that build parameters are declared

### LINKED ISSUE #22: Add validation that parameters are declared
Add a validation that the parameters used in the message and mitigation are declared.

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #22

### CODE REVIEW

File: src/main/java/com/exasol/errorcodecrawlermavenplugin/validation/ParametersAreDeclaredValidator.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import com.exasol.errorcodecrawlermavenplugin.model.ErrorMessageDeclaration;
+import com.exasol.errorcodecrawlermavenplugin.model.NamedParameter;
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * This {@link ErrorMessageDeclarationValidator} validates that all parameters used in message and mitigation are
+ * declared.
+ */
+class ParametersAreDeclaredValidator implements ErrorMessageDeclarationValidator {
+
```
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/errorcodecrawlermavenplugin/validation/ParametersAreDeclaredValidator.java:20]:
```suggestion
```

File: src/main/java/com/exasol/errorcodecrawlermavenplugin/validation/ParametersAreDeclaredValidator.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import com.exasol.errorcodecrawlermavenplugin.Finding;
+import com.exasol.errorcodecrawlermavenplugin.model.ErrorMessageDeclaration;
+import com.exasol.errorcodecrawlermavenplugin.model.NamedParameter;
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * This {@link ErrorMessageDeclarationValidator} validates that all parameters used in message and mitigation are
+ * declared.
+ */
+class ParametersAreDeclaredValidator implements ErrorMessageDeclarationValidator {
```
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/errorcodecrawlermavenplugin/validation/ParametersAreDeclaredValidator.java:19]:
```suggestion
class ErrorMessageParametersValidator implements ErrorMessageDeclarationValidator {
```

File: src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithConcatenatedMessage.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package com.exasol.errorcodecrawlermavenplugin.examples;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * Valid example that is crawled in the tests.
+ */
+public class TestWithConcatenatedMessage {
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithConcatenatedMessage.java:9]:
```suggestion
```

File: src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithMitigations.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package com.exasol.errorcodecrawlermavenplugin.examples;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * Valid example that is crawled in the tests.
+ */
+public class TestWithMitigations {
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithMitigations.java:9]:
```suggestion
```

File: src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithMultiplePartConcatenatedMessage.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.exasol.errorcodecrawlermavenplugin.examples;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * Valid example that is crawled in the tests.
+ */
+public class TestWithMultiplePartConcatenatedMessage {
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithMultiplePartConcatenatedMessage.java:9]:
```suggestion
```

File: src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithNamedParameter.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.exasol.errorcodecrawlermavenplugin.examples;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * Valid example that is crawled in the tests.
+ */
+public class TestWithNamedParameter {
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithNamedParameter.java:9]:
```suggestion
```

File: src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithNamedParameterWithDescription.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package com.exasol.errorcodecrawlermavenplugin.examples;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * Valid example that is crawled in the tests.
+ */
+public class TestWithNamedParameterWithDescription {
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithNamedParameterWithDescription.java:9]:
```suggestion
```

File: src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithNamedUnquotedParameter.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package com.exasol.errorcodecrawlermavenplugin.examples;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * Valid example that is crawled in the tests.
+ */
+public class TestWithNamedUnquotedParameter {
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithNamedUnquotedParameter.java:9]:
```suggestion
```

File: src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithNamedUnquotedParameterWithDescription.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package com.exasol.errorcodecrawlermavenplugin.examples;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * Valid example that is crawled in the tests.
+ */
+public class TestWithNamedUnquotedParameterWithDescription {
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithNamedUnquotedParameterWithDescription.java:9]:
```suggestion
```

File: src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithTwoMessageCalls.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package com.exasol.errorcodecrawlermavenplugin.examples;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * Valid example that is crawled in the tests.
+ */
+public class TestWithTwoMessageCalls {
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithTwoMessageCalls.java:9]:
```suggestion
```

File: src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithUndeclaredParameter.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
+package com.exasol.errorcodecrawlermavenplugin.examples;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * Valid example that is crawled in the tests.
+ */
+public class TestWithUndeclaredParameter {
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/errorcodecrawlermavenplugin/examples/TestWithUndeclaredParameter.java:9]:
```suggestion
```

File: src/test/java/com/exasol/errorcodecrawlermavenplugin/validation/ParametersAreDeclaredValidatorTest.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.junit.jupiter.api.Test;
+
+import com.exasol.errorcodecrawlermavenplugin.Finding;
+import com.exasol.errorcodecrawlermavenplugin.model.ErrorMessageDeclaration;
+
+class ParametersAreDeclaredValidatorTest {
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/errorcodecrawlermavenplugin/validation/ParametersAreDeclaredValidatorTest.java:15]:
```suggestion
```

File: src/test/java/com/exasol/errorcodecrawlermavenplugin/validation/ParametersAreDeclaredValidatorTest.java [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.junit.jupiter.api.Test;
+
+import com.exasol.errorcodecrawlermavenplugin.Finding;
+import com.exasol.errorcodecrawlermavenplugin.model.ErrorMessageDeclaration;
+
+class ParametersAreDeclaredValidatorTest {
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/errorcodecrawlermavenplugin/validation/ParametersAreDeclaredValidatorTest.java:14]:
```suggestion
class ErrorMessageParametersValidatorTest {
```

File: src/main/java/com/exasol/errorcodecrawlermavenplugin/ErrorMessageDeclarationCrawler.java [hunk 14]
```diff
@@ -59,14 +58,22 @@ public ErrorMessageDeclarationCrawler(final Path projectDirectory, final String[
...
-        final List<ErrorMessageDeclaration> errorMessageDeclarations = new ArrayList<>();
-        final SpoonAPI spoon = initSpoon(pathsToCrawl);
-        for (final CtInvocation<?> methodInvocation : spoon.getModel().getRootPackage()
-                .getElements(new TypeFilter<>(CtInvocation.class))) {
-            crawl(methodInvocation, findings, errorMessageDeclarations);
+        try {
+            final List<Finding> findings = new LinkedList<>();
+            final List<ErrorMessageDeclaration> errorMessageDeclarations = new ArrayList<>();
+            final SpoonAPI spoon = initSpoon(pathsToCrawl);
+            for (final CtInvocation<?> methodInvocation : spoon.getModel().getRootPackage()
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/errorcodecrawlermavenplugin/ErrorMessageDeclarationCrawler.java:65]: i tried to clarify.,

File: src/main/java/com/exasol/errorcodecrawlermavenplugin/ErrorMessageDeclarationCrawler.java [hunk 15]
```diff
@@ -168,9 +175,27 @@ private void addBuilderStep(final CtInvocation<?> builderCall,
...
+        } else if (declaringTypeName.equals(ERROR_MESSAGE_BUILDER)
+                && methodSignature.equals("parameter(java.lang.String,java.lang.Object)")) {
+            readParameterStepWithoutDescription(builderCall, errorCodeBuilder, true);
+        } else if (declaringTypeName.equals(ERROR_MESSAGE_BUILDER)
+                && methodSignature.equals("unquotedParameter(java.lang.String,java.lang.Object)")) {
+            readParameterStepWithoutDescription(builderCall, errorCodeBuilder, false);
+        } else if (declaringTypeName.equals(ERROR_MESSAGE_BUILDER)
+                && methodSignature.equals("parameter(java.lang.String,java.lang.Object,java.lang.String)")) {
+            readParameterStepWithDescription(builderCall, errorCodeBuilder, true);
+        } else if (declaringTypeName.equals(ERROR_MESSAGE_BUILDER)
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/errorcodecrawlermavenplugin/ErrorMessageDeclarationCrawler.java:196]: i refactored it.

**Review by chiaradiamarcelo — APPROVED** [review]

**Review by chiaradiamarcelo — APPROVED** [review]

**Pull request merged**